aptos-move/framework/aptos-framework/sources/stake.spec.move

```diff
@@ -162,6 +215,84 @@ spec aptos_framework::stake {
         aborts_if !exists<StakePool>(pool_address);
         aborts_if !exists<ValidatorConfig>(pool_address);
         aborts_if global<ValidatorConfig>(pool_address).validator_index >= len(validator_perf.validators);
+
+        let stake_pool = global<StakePool>(pool_address);
+        let validator_config = global<ValidatorConfig>(pool_address);
+        let cur_validator_perf = validator_perf.validators[validator_config.validator_index];
+        let num_successful_proposals = cur_validator_perf.successful_proposals;
+        let num_total_proposals = cur_validator_perf.successful_proposals + cur_validator_perf.failed_proposals;
+        let rewards_rate = spec_get_reward_rate_1(staking_config);
+        let rewards_rate_denominator = spec_get_reward_rate_2(staking_config);
+        let rewards_amount_1 = if (stake_pool.active.value > 0) {
+            spec_rewards_amount(stake_pool.active.value, num_successful_proposals, num_total_proposals, rewards_rate, rewards_rate_denominator)
+        } else {
+            0
+        };
+        let rewards_amount_2 = if (stake_pool.pending_inactive.value > 0) {
+            spec_rewards_amount(stake_pool.pending_inactive.value, num_successful_proposals, num_total_proposals, rewards_rate, rewards_rate_denominator)
+        } else {
+            0
+        };
+        let post post_stake_pool = global<StakePool>(pool_address);
+        let post post_active_value = post_stake_pool.active.value;
+        let post post_pending_inactive_value = post_stake_pool.pending_inactive.value;
+        let fees_table = global<ValidatorFees>(@aptos_framework).fees_table;
+        let post post_fees_table = global<ValidatorFees>(@aptos_framework).fees_table;
+        let post post_inactive_value = post_stake_pool.inactive.value;
+        ensures post_stake_pool.pending_active.value == 0;
+        ensures if (features::spec_is_enabled(features::COLLECT_AND_DISTRIBUTE_GAS_FEES) && table::spec_contains(fees_table, pool_address)) {
+            !table::spec_contains(post_fees_table, pool_address) &&
+            post_active_value == stake_pool.active.value + rewards_amount_1 + stake_pool.pending_active.value + table::spec_get(fees_table, pool_address).value
+        } else {
+            post_active_value == stake_pool.active.value + rewards_amount_1 + stake_pool.pending_active.value
+        };
+        ensures if (timestamp::spec_now_seconds() >= stake_pool.locked_until_secs) {
+            post_pending_inactive_value == 0 &&
+            post_inactive_value == stake_pool.inactive.value + stake_pool.pending_inactive.value + rewards_amount_2
+        } else {
+            post_pending_inactive_value == stake_pool.pending_inactive.value + rewards_amount_2
+        };
+    }
+
+    // We cannot add helper function in staking_config, so we write two helper functions here to describe function staking_config::get_reward_rate().
```

Comment:
Why couldn't you add helper functions in staking_config?

Reply:
We cannot add helper function in staking_config, cause using pragma opaque in staking_config::get_reward_rate() will cause unexpected error. So we write two helper functions here to fetch the result of staking_config::get_reward_rate().